### DISCUSSION TITLE
Ability to add an attribute to roots.

### DISCUSSION THREAD

**Discussion by cNoNim** [DevTeam/Pure.DI#67] [Ideas]
In #63 assumes that if a composition has the [Module] attribute, then the [Provides] attribute is generated for its roots.

But controlling which attributes are added to roots would be ideal, similar to controlling modifiers via RootKinds.
I'm not sure which API is exactly realizable in terms of source generator capabilities.

Maybe something like:
```csharp
	DI.Setup()
	  .Root<IService>(
		   "Service",
		   attributes:
		   [
			   new ProvidesAttribute(),
			   new MethodImplAttribute(MethodImplOptions.AggressiveInlining)
		   ]);
```

Note: Another RootKinds.Provides for the Module API may be a better choice.
This API will be used a lot.

**Comment by NikolayPianikov** [comment]
@cNoNim It is now possible to create partial methods. See [this example](https://github.com/DevTeam/Pure.DI/blob/master/readme/composition-root-kinds.md). They allow you to define attributes for the roots of a composition. I don't see the point of overloading methods to define roots with an additional argument - a list of attributes.

**Reply by cNoNim** [reply]
That's a good point. No further questions.
Maybe the partial method is better than the RootKinds. Can the generator not add modifiers and add partial automatically to the generated root, if there is a partial method?

**Comment by NikolayPianikov** [comment]
I'll move this ticket to discussion for now for further consideration.
